### PR TITLE
fix(ngResource): fix query url params encoding

### PR DESCRIPTION
**Pull request by alediaferia**
Encoding URL Query Parameters should be "more aggressive" than the URL Segment encoding.
You can see what I mean in the added test case.

I'm currently using AngularJS 1.3.16 but I found this bug is still there in master.

[Review on Reviewable](https://reviewable.io/reviews/angular/angular.js/12201)

### CODE REVIEW

**Comment by alediaferia** [comment]
I signed it!

**Comment by nacmacfeegle** [comment]
+1 to this fix, thanks @alediaferia

File: src/ngResource/resource.js [hunk 1]
```diff
@@ -431,16 +431,20 @@ angular.module('ngResource', ['ng']).
             }
             if (!(new RegExp("^\\d+$").test(param)) && param &&
               (new RegExp("(^|[^\\\\]):" + param + "(\\W|$)").test(url))) {
-              urlParams[param] = true;
+              urlParams[param] = { isQueryParam: (new RegExp("\\?(.*)\\:" + param)).test(url) };
```
Comment (gkalpak) [inline src/ngResource/resource.js]: You need the `'(\\W|$)'` part or else you might get false positive. (Better yet `'(?:\\W|$)'`.)
E.g. `param: 'foo'` and `url: '/path/foo?key=:foobar'` will lead to a false positive.

Comment (alediaferia) [inline src/ngResource/resource.js]: I agree my regex is too broad but I'm not sure I completely get the one proposed by you.
Would something like the following be more accurate?
`new RegExp("\\?((\\w+)\\=\\w+&)*\\w+=:" + param + "(\\W|$)").test(url)`

Comment (gkalpak) [inline src/ngResource/resource.js]: No need to make it too strict either :)

IMO, `new RegExp('\\?(.*):' + param + '(\\W|$)')` (or `new RegExp('\\?(?:.*):' + param + '(?:\\W|$)')`) is fine.

Comment (alediaferia) [inline src/ngResource/resource.js]: So what does `?:` match exactly? This is the only bit I don't understand. :)

Comment (alediaferia) [inline src/ngResource/resource.js]: Ooh, it's the non-capturing group! :)

Comment (alediaferia) [inline src/ngResource/resource.js]: I've updated the code accordingly and split my commit messages to better comply with contribution guidelines.

**Comment by alediaferia** [comment]
I've updated the code accordingly and split my commit messages to better comply with contribution guidelines.

File: src/ngResource/resource.js [hunk 2]
```diff
@@ -431,16 +431,20 @@ angular.module('ngResource', ['ng']).
             }
             if (!(new RegExp("^\\d+$").test(param)) && param &&
               (new RegExp("(^|[^\\\\]):" + param + "(\\W|$)").test(url))) {
-              urlParams[param] = true;
+              urlParams[param] = { isQueryParam: (new RegExp("\\?(?:.*)\\:" + param + "(?:\\W|$)")).test(url) };
```
Comment (gkalpak) [inline src/ngResource/resource.js]: Nitpick: You don't need `\\` before `:` (it's not a special character in this context).

**Comment by gkalpak** [comment]
LGTM

**Comment by davidesantangelo** [comment]
LGMT

**Comment by alediaferia** [comment]
Is there a plan for when this will go in the main line?

**Comment by alediaferia** [comment]
Bump!

**Comment by alediaferia** [comment]
Is there something I can do to ease the merge for this patch?

**Comment by gkalpak** [comment]
@alediaferia, sorry it wasn't brought up earlier (I didn't spot it originally - thx @petebacondarwin for pointing it out), but this seems to be a breaking change.
E.g., one might want to specify the whole query string as param, like this:

```js
$resource('/api/myapp/resource?:query').get({query: 'foo&bar'});
```

--
So, in order to avoid the breaking change and still benefit from the fix, I propose that the `&` is only encoded, if the param is a recognized as query string value part (of a key=value pair).

This can be achieved, with the following changes:

1. Change this line:
        `urlParams[param] = { isQueryParam: (new RegExp("\\?(?:.*):" + param + "(?:\\W|$)")).test(url) };`
    to that:
        `urlParams[param] = { isQueryParamValue: (new RegExp("\\?(?:.*)=:" + param + "(?:\\W|$)")).test(url) };`

2. Change `paramInfo.isQueryParam` to `paramInfo.isQueryParamValue` (or whatever you name it).

3. Modify the test, in order to check for both cases:

        it('should encode & in query params unless in query param value', function() {
          var R1 = $resource('/api/myapp/resource?:query');
          $httpBackend.expect('GET', '/api/myapp/resource?foo&bar').respond({});
          R1.get({query: 'foo&bar'});
          
          var R2 = $resource('/api/myapp/resource?from=:from');
          $httpBackend.expect('GET', '/api/myapp/resource?from=bar%20%26%20blanks').respond({});
          R2.get({from: 'bar & blanks'});
        });

--
@alediaferia, would you be up to making those changes and see what people think ?

**Comment by alediaferia** [comment]
@gkalpak will change according to your suggestions ASAP. Thank you for getting back :)

**Comment by petebacondarwin** [comment]
@alediaferia nice one. Keep your ki cool. Thanks for the help.

File: src/ngResource/resource.js [hunk 3]
```diff
@@ -433,7 +433,7 @@ angular.module('ngResource', ['ng']).
             }
             if (!(new RegExp("^\\d+$").test(param)) && param &&
               (new RegExp("(^|[^\\\\]):" + param + "(\\W|$)").test(url))) {
-              urlParams[param] = true;
+              urlParams[param] = { isQueryParamValue: (new RegExp("\\?(?:.*)=:" + param + "(?:\\W|$)")).test(url) };
```
Comment (gkalpak) [inline src/ngResource/resource.js]: Hm...looking at it again, why not change `(?:.*)` to `.*` ?

File: test/ngResource/resourceSpec.js [hunk 4]
```diff
@@ -338,6 +338,17 @@ describe("resource", function() {
   });
 
 
+  it('should handle url encoding in mapped params', function() {
```
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: I'd prefer a more descriptive testcase name (e.g. `should encode & in query params unless in query param value` or something).

File: src/ngResource/resource.js (hunk 3 above)
Comment (alediaferia) [inline src/ngResource/resource.js]: Makes sense

File: src/ngResource/resource.js [hunk 5]
```diff
@@ -443,10 +443,14 @@ angular.module('ngResource', ['ng']).
           });
 
           params = params || {};
-          forEach(self.urlParams, function(_, urlParam) {
+          forEach(self.urlParams, function(paramInfo, urlParam) {
             val = params.hasOwnProperty(urlParam) ? params[urlParam] : self.defaults[urlParam];
             if (angular.isDefined(val) && val !== null) {
-              encodedVal = encodeUriSegment(val);
+              if (paramInfo.isQueryParamiValue === true) {
```
Comment (gkalpak) [inline src/ngResource/resource.js]: _isQueryParamiValue_  -->  _isQueryParamValue_

(It's a good idea to locally run the tests before pushing, to ensure nothing is (obviously) breaking :smiley:)

**Comment by gkalpak** [comment]
@alediaferia, thx for the update. I left a couple of comments (mostly minor).
Feel free to ping me once you've pushed changes, so I can take another look.

File: src/ngResource/resource.js (hunk 5 above)
Comment (alediaferia) [inline src/ngResource/resource.js]: I'm sorry for that. Actually I thought I pushed just to my `origin` and not to `upstream` as well. Still didn't manage to run the tests. Will do ASAP

Comment (gkalpak) [inline src/ngResource/resource.js]: NP :)

BTW, when you have submitted a PR, you don't need to push upstream or anything; pushing to your origin/<branch> will auto-update the PR (GitHub feature), so that's probably what happened.

Just ping me when you're done, so I take a final look. Thx for all the work so far !

**Comment by alediaferia** [comment]
@gkalpak BTW do you have any hint on why now my `grunt package` fails with `stderr maxBuffer exceeded` ?

**Comment by gkalpak** [comment]
@alediaferia, no idea :) Does it happen with the code currently on the PR ?
(If not, you can push the code somewhere so I can take a look.)

**Comment by petebacondarwin** [comment]
Sounds like something to do with the closure compiler?

**Comment by alediaferia** [comment]
@gkalpak @petebacondarwin sorry for not getting back earlier. It's happening with the current PR but it's happening with updated master as well. It may be something wrong with my npm/node configuration, but not sure. I'm not much into JavaScript :)

**Comment by gkalpak** [comment]
@alediaferia, (un)fortunately, I'm not getting any issues with `grunt package` :disappointed:

File: test/ngResource/resourceSpec.js [hunk 6]
```diff
@@ -338,6 +338,17 @@ describe("resource", function() {
   });
 
 
+  it('should encode & in query params unless in query param value', function() {
```
Comment (alediaferia) [inline test/ngResource/resourceSpec.js]: I'm wondering if part of this test is a duplicate of the one preceding this test.

**Comment by alediaferia** [comment]
@gkalpak ping

File: test/ngResource/resourceSpec.js (hunk 6 above)
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: I would say close but not quite.
I would be happy to have them merged into one `it` block (keeping all 3 cases).

Feel free to merge them.

File: scripts/npm/install-dependencies.sh [hunk 7]
```diff
@@ -10,7 +10,7 @@ if diff -q $SHRINKWRAP_FILE $SHRINKWRAP_CACHED_FILE; then
 else
   echo 'Blowing away node_modules and reinstalling npm dependencies...'
   rm -rf node_modules
-  npm install
+  npm install --registry http://registry.npmjs.org/
```
Comment (gkalpak) [inline scripts/npm/install-dependencies.sh]: I guess this has to do with your `grunt package` issues, but it definitely does not belong in this PR :)

Comment (alediaferia) [inline scripts/npm/install-dependencies.sh]: Woops! sorry :)

**Comment by gkalpak** [comment]
LGTM, thx @alediaferia !

/cc @petebacondarwin

**Comment by alediaferia** [comment]
If this is ok, shall I clean up the git history squashing the latest commits and following your commit message conventions?

**Comment by petebacondarwin** [comment]
OK, LGTM. @alediaferia thanks for your work on this. If you want to squash and tidy up the commit message that would be great. Otherwise, perhaps @gkalpak could do it when merging?

**Comment by gkalpak** [comment]
Sure, I'll take care of squashing and merging !

**Comment by alediaferia** [comment]
Thank you @gkalpak @petebacondarwin

**Comment by gkalpak** [comment]
@petebacondarwin, should this be backported to 1.4.x ?

**Comment by petebacondarwin** [comment]
If you are confident it is not a breaking change.

**Comment by gkalpak** [comment]
It's one of those times that I find it hard to distinguish between a fix and a BC.

The only change introduced is when a param (`:xyz`) is used as the _value part_ of a URL query key-value pair. In any other case, there is no difference. But it could still break an app (theoretically).

E.g., assuming `$resource('/path?key=:val').get({val: 'foo&bar'})`, the request is sent to this URL:

1.4.x: `/path?key=foo&bar`
1.5.x: `/path?key=foo%26bar`

**Comment by alediaferia** [comment]
Although I feel the "new" behaviour is the right one I unfortunately think there may be a lot of apps out there relying on a workaround for this "issue". 

If you look for e.g. `"angularjs ngresource query parameters"` on Google there are some issues open on SO with a few workarounds that usually involve using `$http` directly.

Just my 2 cents.

Happy to have contributed to such a big project anyway :)

**Comment by petebacondarwin** [comment]
OK, so no backport.

**Comment by gkalpak** [comment]
@alediaferia, I couldn't find any SO questions/answers that would be broken by this "fix". Relying on a workaround isn't a problem (it won't get broken).
Do you have any specific SO question/answer in mind ?

The problem is if someone has as a URL template such as `/path?key=:val` and expects to set other query param via `:val` (e.g. `/path?key=1&key2=2&key3=3`). It is a weird usecase imo (but theoretically possible nontheless).

**Comment by alediaferia** [comment]
@gkalpak I cannot manage to find the question I had in mind, strange. Anyway, as you point out, relying on a workaround won't make things break. That's true. I don't have further objections.
